multi_partner_learning.py

```diff
@@ -302,6 +315,112 @@ def compute_collaborative_round_fedavg(self):
 
         logger.debug("End of fedavg collaborative round.")
 
+
+    def compute_collaborative_round_qavg(self):
+            """Proceed to a collaborative round with a quality aware federated averaging approach"""
+
+            logger.debug("Start new qavg collaborative round ...")
+
+            # Initialize variables
+            epoch_index, minibatch_index = self.epoch_index, self.minibatch_index
+            is_very_first_minibatch = (epoch_index == 0 and minibatch_index == 0)
+            x_val, y_val = self.val_data
+            partners_list = self.partners_list
+
+            # Starting model for each partner is the aggregated model from the previous mini-batch iteration
+            if is_very_first_minibatch:  # Except for the very first mini-batch where it is a new model
+                logger.debug(f"(fedavg) Very first minibatch of epoch n°{epoch_index}, init new models for each partner")
+                partners_model_list_for_iteration = self.init_with_new_models()
+            else:
+                logger.debug(f"(fedavg) Minibatch n°{minibatch_index} of epoch n°{epoch_index}, "
+                            f"init aggregated model for each partner with models from previous round")
+                partners_model_list_for_iteration = self.init_with_agg_models()
+
+            # model_to_evaluate = partners_model_list_for_iteration[0]
+            # model_evaluation = model_to_evaluate.evaluate(x_val, y_val, batch_size=constants.DEFAULT_BATCH_SIZE, verbose=0)
+            # self.score_matrix_collective_models[epoch_index, minibatch_index] = model_evaluation[1]
```

Comment:
By commenting these 3 lines you remove the evaluation of the aggregated model that is used at the start of this collaborative round. So in the `score_matrix_collective_models` you will not have a performance value.

I understand that in the code below you develop an evaluation approach on test data from *other partners*. But you apply it only on models trained individually on partners, and this goes in `score_matrix_per_partner`. To be complete you should consider also evaluating the aggregated model at the beginning of the collaborative round (either with the global val data or with your special evaluation approach).

Reply:
It looks like it is now fixed, but I am not 100% sure, I'll leave it to you @MrDaiki to comment about it